src/engine/plugins/match.cpp

```diff
@@ -218,6 +228,27 @@ Status MatchPlugin::HandleRequest(const RoutingAlgorithmsInterface &algorithms,
         return Error("NoMatch", "Could not match the trace.", json_result);
     }
 
+    // trace was split, we don't support the waypoints parameter across multiple match objects
+    if (sub_matchings.size() > 1 && !parameters.waypoints.empty())
+    {
+        return Error("NoMatch", "Could not match the trace with the given waypoints.", json_result);
+    }
+
+    // Error: Check if user-supplied waypoints can be found in the resulting matches
+    for (const auto waypoint : tidied.parameters.waypoints)
```

Comment:
The complexity of this check is too high (three nested loops). What about using a `std::set<std::size_t>' that contains the waypoint indices. Then we could loop over all sub-matching indices and remove the index of the waypoint we found.